pipenv/vendor/pythonfinder/models/python.py

```diff
@@ -318,50 +318,80 @@ def which(self, name) -> PathEntry | None:
+                                    version = self._get_registry_value(tag_key, "Version", default=tag[:3])
+                                    sys_version = self._get_registry_value(tag_key, "SysVersion", default=tag[:3])
+                                    sys_architecture = self._get_registry_value(tag_key, "SysArchitecture")
+
+                                    if company == "PythonCore" and not sys_architecture:
+                                        # TODO: Implement PEP-514 defaults for architecture for PythonCore based on key and OS architecture.
+                                        sys_architecture = None
+
+                                    try:
+                                        with winreg.OpenKey(tag_key, "InstallPath") as install_path_key:
+                                            install_path = self._get_registry_value(install_path_key, None)
+                                            executable_path = self._get_registry_value(install_path_key,
+                                                                                       "ExecutablePath")
+                                            windowed_executable_path = self._get_registry_value(install_path_key,
+                                                                                                "WindowedExecutablePath")
+
+                                            if company == "PythonCore" and not executable_path:
+                                                executable_path = os.path.join(install_path, "python.exe")
+                                            if company == "PythonCore" and not windowed_executable_path:
+                                                windowed_executable_path = os.path.join(install_path, "pythonw.exe")
+
+                                            launcher_entry = WindowsLauncherEntry(
+                                                version=Version(sys_version),
+                                                executable_path=executable_path,
+                                                windowed_executable_path=windowed_executable_path,
+                                                company=company,
+                                                tag=tag,
+                                                display_name=display_name,
+                                                support_url=support_url,
+                                                architecture=sys_architecture,
+                                            )
+                                            yield launcher_entry
+
+                                    except FileNotFoundError:
+                                        continue
+
+            except FileNotFoundError:
+                pass
+
+    def _get_registry_value(self, key, value_name, default=None):
+        try:
+            import winreg
+            return winreg.QueryValue(key, value_name)
```

Comment:
Although this matches the code from PEP-514, it doesn't actually work. (I just checked, and the example from PEP-514 also fails. With the PEP-514 example code, my system-wide 64-bit Python 3.10 install reports as "Python 3.1 (Unknown)". Because of the default logic, at the time of PEP-514's writing, the output was mostly correct, so maybe no one noticed?)

It should be something like:

```suggestion
            import winreg
            # TODO: Not specified in PEP-514, but probably should reject any value that isn't `winreg.REG_SZ`.
            # PEP-514's (faulty) example code uses `winreg.QueryValue`, which can only return `winreg.REG_SZ` values.
            # They document "string value" which could match `winreg.REG_SZ` or `winreg.REG_EXPAND_SZ`.
            return winreg.QueryValueEx(key, value_name)[0]
```

This is because `winreg.QueryValue`'s second parameter is actually a sub-key to query for its default value, not a value name. `winreg.QueryValueEx` expects you to have opened the target key already (as we have) and so its second parameter is the value name, or `None` for the default value.

I note that pep514tools doesn't use `winreg.QueryValue` either, in favour of dumping all the values into a dictionary using `winreg.EnumValue`, so it wouldn't have seen this issue. It also only accepts `winreg.REG_SZ` values.